File: matching-engine/src/test/java/com/titanx/risk/RiskEngineTest.java
package com.titanx.risk;

import com.titanx.model.*;
import org.junit.jupiter.api.*;

import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.*;

@DisplayName("TitanX RiskEngine — Pre-Trade Check Tests")
class RiskEngineTest {

    private RiskEngine riskEngine;
    private CircuitBreaker circuitBreaker;

    @BeforeEach
    void setUp() {
        circuitBreaker = new CircuitBreaker();
        riskEngine = new RiskEngine(circuitBreaker);
    }

    @Test
    @DisplayName("Valid order passes all risk checks")
    void validOrderIsApproved() {
        Order order = Order.create("acc-001", "AAPL", OrderSide.BUY, OrderType.LIMIT,
                new BigDecimal("150.00"), null, 100);
        RiskResult result = riskEngine.checkPreTrade(order);
        assertThat(result.approved()).isTrue();
    }

    @Test
    @DisplayName("Order exceeding notional limit is rejected")
    void notionalLimitRejected() {
        // price * qty = 600.00 * 10000 = $6M > $5M limit
        Order bigOrder = Order.create("acc-001", "AAPL", OrderSide.BUY, OrderType.LIMIT,
                new BigDecimal("600.00"), null, 10_000);
        RiskResult result = riskEngine.checkPreTrade(bigOrder);
        assertThat(result.approved()).isFalse();
        assertThat(result.reason()).isEqualTo("NOTIONAL_LIMIT_EXCEEDED");
    }

    @Test
    @DisplayName("Duplicate order ID is rejected")
    void duplicateOrderIdRejected() {
        Order order = Order.create("acc-001", "AAPL", OrderSide.BUY, OrderType.LIMIT,
                new BigDecimal("100.00"), null, 50);
        riskEngine.checkPreTrade(order);  // first submission — accepted

        // Simulate replay with same order object (same orderId)
        RiskResult duplicate = riskEngine.checkPreTrade(order);
        assertThat(duplicate.approved()).isFalse();
        assertThat(duplicate.reason()).isEqualTo("DUPLICATE_ORDER_ID");
    }

    @Test
    @DisplayName("Order rejected when circuit breaker is triggered")
    void circuitBreakerRejectsOrder() {
        circuitBreaker.haltSymbol("AAPL", "Test halt");

        Order order = Order.create("acc-001", "AAPL", OrderSide.BUY, OrderType.LIMIT,
                new BigDecimal("100.00"), null, 100);
        RiskResult result = riskEngine.checkPreTrade(order);

        assertThat(result.approved()).isFalse();
        assertThat(result.reason()).isEqualTo("CIRCUIT_BREAKER_HALT");
    }

    @Test
    @DisplayName("Position limit exceeded after accumulated fills")
    void positionLimitExceeded() {
        // Build up a position via recordFill
        riskEngine.recordFill("acc-001", "AAPL", OrderSide.BUY, 9_500);

        // Now try to buy 1000 more — would take position to 10,500 > 10,000 limit
        Order order = Order.create("acc-001", "AAPL", OrderSide.BUY, OrderType.LIMIT,
                new BigDecimal("100.00"), null, 1_000);
        RiskResult result = riskEngine.checkPreTrade(order);

        assertThat(result.approved()).isFalse();
        assertThat(result.reason()).isEqualTo("POSITION_LIMIT_EXCEEDED");
    }

    @Test
    @DisplayName("Circuit breaker triggers on large price move")
    void circuitBreakerTriggersOnPriceMove() {
        circuitBreaker.recordTrade("TSLA", new BigDecimal("200.00")); // reference price
        // Simulate a 15% crash — exceeds 5% threshold
        circuitBreaker.recordTrade("TSLA", new BigDecimal("169.00"));

        assertThat(circuitBreaker.isHalted("TSLA")).isTrue();
    }

    @Test
    @DisplayName("Circuit breaker resumes after manual clear")
    void circuitBreakerResumes() {
        circuitBreaker.haltSymbol("NVDA", "Test");
        assertThat(circuitBreaker.isHalted("NVDA")).isTrue();

        circuitBreaker.resumeSymbol("NVDA");
        assertThat(circuitBreaker.isHalted("NVDA")).isFalse();
    }
}

File: backend/tests/RiskEngineTest.cpp
#include <gtest/gtest.h>
#include "titanx/risk/RiskEngine.hpp"
#include "titanx/risk/CircuitBreaker.hpp"

using namespace titanx;

/**
 * ============================================================
 * RiskEngine Unit Tests
 * ============================================================
 *
 * Tests cover all 6 pre-trade checks:
 *  1. Circuit breaker halt
 *  2. Duplicate order ID
 *  3. Notional limit ($5M)
 *  4. Position limit (10K shares)
 *  5. Fat-finger protection (10% deviation)
 *  6. Self-trade prevention
 */

class RiskEngineTest : public ::testing::Test {
protected:
    CircuitBreaker circuitBreaker;
    RiskEngine riskEngine{circuitBreaker};
};

// ---- 1. Circuit breaker ----

TEST_F(RiskEngineTest, RejectsWhenCircuitBreakerTriggered) {
    circuitBreaker.haltSymbol("AAPL", "test");

    auto order = Order::create("acc1", "AAPL", OrderSide::BUY,
                               OrderType::LIMIT, price_to_fixed(150.0),
                               0, 100);
    auto result = riskEngine.checkPreTrade(order);

    EXPECT_FALSE(result.approved);
    EXPECT_EQ(result.reason, "CIRCUIT_BREAKER_HALT");
}

// ---- 2. Duplicate order ID ----

TEST_F(RiskEngineTest, RejectsDuplicateOrderId) {
    auto order = Order::create("acc1", "AAPL", OrderSide::BUY,
                               OrderType::LIMIT, price_to_fixed(150.0),
                               0, 100);

    auto result1 = riskEngine.checkPreTrade(order);
    EXPECT_TRUE(result1.approved);

    // Same order submitted again
    auto result2 = riskEngine.checkPreTrade(order);
    EXPECT_FALSE(result2.approved);
    EXPECT_EQ(result2.reason, "DUPLICATE_ORDER_ID");
}

// ---- 3. Notional limit ----

TEST_F(RiskEngineTest, RejectsExcessiveNotional) {
    // $10,000 × 600 shares = $6M notional > $5M limit
    auto order = Order::create("acc1", "AAPL", OrderSide::BUY,
                               OrderType::LIMIT, price_to_fixed(10000.0),
                               0, 600);
    auto result = riskEngine.checkPreTrade(order);

    EXPECT_FALSE(result.approved);
    EXPECT_EQ(result.reason, "NOTIONAL_LIMIT_EXCEEDED");
}

TEST_F(RiskEngineTest, AcceptsValidNotional) {
    // $150 × 100 shares = $15K << $5M
    auto order = Order::create("acc1", "AAPL", OrderSide::BUY,
                               OrderType::LIMIT, price_to_fixed(150.0),
                               0, 100);
    auto result = riskEngine.checkPreTrade(order);
    EXPECT_TRUE(result.approved);
}

// ---- 4. Position limit ----

TEST_F(RiskEngineTest, RejectsPositionExceedingLimit) {
    // Simulate holding 9,900 shares
    riskEngine.recordFill("acc1", "AAPL", OrderSide::BUY, 9900);

    // Trying to buy 200 more → 10,100 > 10,000 limit
    auto order = Order::create("acc1", "AAPL", OrderSide::BUY,
                               OrderType::LIMIT, price_to_fixed(150.0),
                               0, 200);
    auto result = riskEngine.checkPreTrade(order);

    EXPECT_FALSE(result.approved);
    EXPECT_EQ(result.reason, "POSITION_LIMIT_EXCEEDED");
}

// ---- 5. Fat-finger protection ----

TEST_F(RiskEngineTest, RejectsFatFingerPrice) {
    // Set last trade at $150
    riskEngine.updateLastTradePrice("AAPL", price_to_fixed(150.0));

    // Order at $200 = 33% deviation > 10% threshold
    auto order = Order::create("acc1", "AAPL", OrderSide::BUY,
                               OrderType::LIMIT, price_to_fixed(200.0),
                               0, 100);
    auto result = riskEngine.checkPreTrade(order);

    EXPECT_FALSE(result.approved);
    EXPECT_EQ(result.reason, "FAT_FINGER_PRICE_DEVIATION");
}

TEST_F(RiskEngineTest, AcceptsReasonablePriceDeviation) {
    riskEngine.updateLastTradePrice("AAPL", price_to_fixed(150.0));

    // Order at $155 = 3.3% deviation < 10% threshold
    auto order = Order::create("acc1", "AAPL", OrderSide::BUY,
                               OrderType::LIMIT, price_to_fixed(155.0),
                               0, 100);
    auto result = riskEngine.checkPreTrade(order);
    EXPECT_TRUE(result.approved);
}

// ---- 6. Self-trade prevention ----

TEST_F(RiskEngineTest, PreventsSelfTrade) {
    // Submit a buy order from acc1
    auto buyOrder = Order::create("acc1", "AAPL", OrderSide::BUY,
                                  OrderType::LIMIT, price_to_fixed(150.0),
                                  0, 100);
    auto result1 = riskEngine.checkPreTrade(buyOrder);
    EXPECT_TRUE(result1.approved);

    // Now try to submit a sell from same account
    auto sellOrder = Order::create("acc1", "AAPL", OrderSide::SELL,
                                   OrderType::LIMIT, price_to_fixed(155.0),
                                   0, 50);
    auto result2 = riskEngine.checkPreTrade(sellOrder);
    EXPECT_FALSE(result2.approved);
    EXPECT_EQ(result2.reason, "SELF_TRADE_PREVENTION");
}

// ---- Counters ----

TEST_F(RiskEngineTest, TracksCheckAndRejectCounters) {
    auto order1 = Order::create("acc1", "AAPL", OrderSide::BUY,
                                OrderType::LIMIT, price_to_fixed(150.0),
                                0, 100);
    riskEngine.checkPreTrade(order1);

    circuitBreaker.haltSymbol("AAPL", "test");
    auto order2 = Order::create("acc2", "AAPL", OrderSide::BUY,
                                OrderType::LIMIT, price_to_fixed(150.0),
                                0, 100);
    riskEngine.checkPreTrade(order2);

    EXPECT_EQ(riskEngine.totalChecked(), 2);
    EXPECT_EQ(riskEngine.totalRejected(), 1);
}
